File: Data Structures & Algorithms/house-robber/submission-2.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];
        else if (nums.size() == 2) return max(nums[0], nums[1]);

        int prev = nums[0];
        int curr = max(nums[0], nums[1]);

        for (int i = 2; i < nums.size(); ++i) {
            int temp = curr;
            curr = max(nums[i] + prev, curr);
            prev = temp;
        }

        return curr;
    }
};

File: Data Structures & Algorithms/burst-balloons/submission-2.py
from functools import cache

class Solution:
    def maxCoins(self, nums: List[int]) -> int:

        @cache
        def dfs(curr):
            if len(curr) == 1:
                return curr[0]
        
            best = -1
            for i in range(len(curr)):
                a = curr[i - 1] if (i - 1) >= 0 else 1
                b = curr[i + 1] if (i + 1) < len(curr) else 1
                res = a * curr[i] * b

                copy = curr[:i] + curr[i + 1:]

                best = max(best, res + dfs(copy))

            return best
        
        return dfs(tuple(nums))